presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownAggregation.java

```diff
@@ -0,0 +1,118 @@
+                .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+
+        if (pushdownExpressions.isEmpty()) {
+            return Result.empty();
+        }
+
+        ImmutableList.Builder<Expression> expressionBuilder = ImmutableList.builder();
+        for (Aggregation aggregation : aggregationNode.getAggregations().values()) {
+            expressionBuilder.addAll(aggregation.getCall().getArguments());
+        }
+        List<DereferenceExpression> aggregations = expressionBuilder.build().stream()
+                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
+                .collect(toImmutableList());
+
+        Map<Symbol, DereferenceExpression> aggregationDereferences = aggregations.stream()
+                .distinct()
+                .filter(expression -> !expressions.containsKey(expression))
+                .collect(toImmutableMap(expression -> newSymbol(expression, context, metadata, sqlParser), Function.identity()));
+
+        List<Symbol> outputs = outputSymbols.stream()
+                .filter(symbol -> assignments.getMap().containsKey(symbol))
+                .collect(toImmutableList());
+
+        Assignments newAssignments = Assignments.builder()
+                .putAll(pushdownExpressions)
+                .putIdentities(outputs)
+                .putAll(aggregationDereferences)
+                .build();
+        ProjectNode child = new ProjectNode(context.getIdAllocator().getNextId(), aggregationNode.getSource(), newAssignments);
+
+        ImmutableMap.Builder<DereferenceExpression, Symbol> constraintBuilder = ImmutableMap.builder();
+        constraintBuilder.putAll(expressions);
+        constraintBuilder.putAll(aggregationDereferences.entrySet().stream().collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey)));
+        Map<DereferenceExpression, Symbol> constraints = constraintBuilder.build();
+
+        ImmutableMap.Builder<Symbol, Aggregation> aggregationBuilder = ImmutableMap.builder();
+        for (Map.Entry<Symbol, Aggregation> entry : aggregationNode.getAggregations().entrySet()) {
+            Symbol symbol = entry.getKey();
+            Aggregation aggregation = entry.getValue();
+            FunctionCall functionCall = aggregation.getCall();
+            ImmutableList.Builder<Expression> builder = ImmutableList.builder();
+            for (Expression expression : functionCall.getArguments()) {
+                builder.add(replaceDereferences(expression, constraints));
+            }
+            FunctionCall targetCall = new FunctionCall(functionCall.getName(), functionCall.getWindow(), functionCall.getFilter(), functionCall.getOrderBy(), functionCall.isDistinct(), builder.build());
+            Aggregation targetAggregation = new Aggregation(targetCall, aggregation.getFunctionHandle(), aggregation.getMask());
+            aggregationBuilder.put(symbol, targetAggregation);
```

Comment:
It might be easier to read as

```
            Aggregation aggregation = entry.getValue();
            FunctionCall functionCall = aggregation.getCall();
            ImmutableList<Expression> newArguments = functionCall.getArguments().stream()
                    .map(expression -> replaceDereferences(expression, constraints))
                    .collect(toImmutableList());
            Aggregation newAggregation = new Aggregation(
                    new FunctionCall(
                            functionCall.getName(),
                            functionCall.getWindow(),
                            functionCall.getFilter(),
                            functionCall.getOrderBy(),
                            functionCall.isDistinct(),
                            newArguments),
                    aggregation.getFunctionHandle(), 
                    aggregation.getMask());
            aggregationBuilder.put(entry.getKey(), newAggregation);
```